ext/mysqli/mysqli_api.c

```diff
@@ -476,6 +476,148 @@ PHP_FUNCTION(mysqli_stmt_execute)
 }
 /* }}} */
 
+void close_stmt_and_copy_errors(MY_STMT *stmt, MY_MYSQL *mysql)
+{
+	/* mysql_stmt_close() clears errors, so we have to store them temporarily */
+	MYSQLND_ERROR_INFO error_info = *stmt->stmt->data->error_info;
+	stmt->stmt->data->error_info->error_list.head = NULL;
+	stmt->stmt->data->error_info->error_list.tail = NULL;
+	stmt->stmt->data->error_info->error_list.count = 0;
+
+	/* we also remember affected_rows which gets cleared too */
+	uint64_t affected_rows = mysql->mysql->data->upsert_status->affected_rows;
+
+	mysqli_stmt_close(stmt->stmt, false);
+	stmt->stmt = NULL;
+	php_clear_stmt_bind(stmt);
+
+	/* restore error messages, but into the mysql object */
+	zend_llist_clean(&mysql->mysql->data->error_info->error_list);
+	*mysql->mysql->data->error_info = error_info;
+	mysql->mysql->data->upsert_status->affected_rows = affected_rows;
+}
+
+PHP_FUNCTION(mysqli_execute_query)
+{
+	MY_MYSQL		*mysql;
+	MY_STMT			*stmt;
+	char			*query = NULL;
+	size_t				query_len;
+	zval			*mysql_link;
+	HashTable	*input_params = NULL;
+	MYSQL_RES 		*result;
+	MYSQLI_RESOURCE	*mysqli_resource;
+
+	if (zend_parse_method_parameters(ZEND_NUM_ARGS(), getThis(), "Os|h!", &mysql_link, mysqli_link_class_entry, &query, &query_len, &input_params) == FAILURE) {
+		RETURN_THROWS();
+	}
+	MYSQLI_FETCH_RESOURCE_CONN(mysql, mysql_link, MYSQLI_STATUS_VALID);
+
+	stmt = (MY_STMT *)ecalloc(1,sizeof(MY_STMT));
+
+	if (!(stmt->stmt = mysql_stmt_init(mysql->mysql))) {
+		MYSQLI_REPORT_MYSQL_ERROR(mysql->mysql);
+		efree(stmt);
+		RETURN_FALSE;
+	}
+
+	if (FAIL == mysql_stmt_prepare(stmt->stmt, query, query_len)) {
+		MYSQLI_REPORT_STMT_ERROR(stmt->stmt);
+		
+		close_stmt_and_copy_errors(stmt, mysql);
+		RETURN_FALSE;
+	}
+
+	/* The bit below, which is copied from mysqli_stmt_prepare, is needed for bad index exceptions */ 
+	/* don't initialize stmt->query with NULL, we ecalloc()-ed the memory */
+	/* Get performance boost if reporting is switched off */
+	if (query_len && (MyG(report_mode) & MYSQLI_REPORT_INDEX)) {
+		stmt->query = (char *)emalloc(query_len + 1);
```

Comment:
nit: estr(n)dup might be shorter

Reply:
Thanks. I hope I used it correctly. I changed it in both places.

Reply:
Sure even tough maybe `estrndup` would have been slightly more appropriate since you have the length already but that s detail :)